### ISSUE TITLE
Performance research

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#5880]
Let's start from creating a set of manual tests on which we'll be able to reliably and quickly test typical performance-oriented scenarios.

Some ideas for those scenarios:

*   Loading large HTML data with a lot of attributes.
    *   Coming from an editor (so, only elements and structures that the editor could've produced).
    *   Coming from external sources (e.g. other website), so with structures, elements and attributes that the editor doesn't understand and many inline styles (so we can test the new style normalization).
*   An editor with a long content (same cases as above) but focused on testing features inside the editor (so the content should be already preloaded).
*   Inserting content (like pasting) into an editor.
    *   Also, 3-4 data sets.
    *   The editor content should be diverse as well so you can test various scenarios (pasting into a table, image caption, middle of something, etc.)
*   Loading a couple of editors at once. We expect to see a linear time growth, so we could test just one, but let's have ~3-5 as at those points the load times start to be noticeable.

Other considerations:

*   The action to be tested should be triggered by a button (instead of happening automatically).
*   Use rich setups of editor plugins (so we cover diverse types of those).

Subtasks:

- [x] Test with empty editor and `setData` button (https://github.com/ckeditor/ckeditor5/pull/6103)
- [x] Test with content initialization (after pressing a button) (#6124)
- [ ] Generate content (small, medium, large)
    - [ ] Base text (content that will serve as a reference for other small, medium, large variations) (https://github.com/ckeditor/ckeditor5/pull/6147)
    - [ ] Attributes known to the editor
    - [ ] Attributes unknown to the editor
    - [ ] Inline, combined CSS styles (https://github.com/ckeditor/ckeditor5/pull/6148)
- [x] Editor with many features enabled (content preloaded) (https://github.com/ckeditor/ckeditor5/pull/6149)
- [ ] Loading multiple editors at once (https://github.com/ckeditor/ckeditor5/pull/6144)
- [x] Review performance, check for any potential bottlenecks (https://github.com/ckeditor/ckeditor5/issues/5880#issuecomment-581922363)

**Comment by zadam** [comment]
It's great to see effort on this topic. As an example I'm attaching a HTML file which takes couple of seconds (5-10 on my machine) to load into CKEditor. The file is big-ish, but still something you might expect to work with in the CKEditor.

[BMesh Operators.zip](https://github.com/ckeditor/ckeditor5/files/4014457/BMesh.Operators.zip)

**Comment by mlewand** [comment]
Added subtasks list for better tracking.

**Comment by Reinmar** [comment]
One note – I saw the newly added tests. They use `fetch()` in `getFileContents()`. AFAIK you can do a simple `import 'path' as foo` or something similar. IIRC we have a loader for txt files. You can check how it works in PFO tests.

**Comment by mlewand** [comment]
tl;dr; for small editors most of the time goes to editor UI rendering (e.g. color dropdowns, table dropdown). In case of editors with a longer content we'd need to optimize data processing.

## [editorinit](http://localhost:8125/ckeditor5/tests/manual/performance/editorinit.html) - short, semantic content

![Performance chart - small semantic content](https://i.imgur.com/IWszUou.png)

![Bottoms up - small semantic content](https://i.imgur.com/fDXCiuC.png)

Total click handler time 1.14s.

Most of the time is consumed by initializing the toolbar (582ms), with font bg/color dropdown taking the longest time to render:

![](https://user-images.githubusercontent.com/5353898/73747084-eb613f80-4756-11ea-9f58-6c7e7099d10c.png)

Which is similar to #6161.

## [editorinit](http://localhost:8125/ckeditor5/tests/manual/performance/editorinit.html) - medium, semantic content

![](https://user-images.githubusercontent.com/5353898/73747308-77736700-4757-11ea-9471-00deb2fe51db.png)

![](https://user-images.githubusercontent.com/5353898/73747402-a8ec3280-4757-11ea-9427-6faf40bebf77.png)

As editor contains more content, it's no longer the UI - instead lodash's `getRawTag` and `src/model/position#get parent` function starts to be visible.

Taking a step back and looking at the bigger picture, it's `_runPendingChanges` where the 85% (both calls combined) of total time goes, so this is where we should look for improvements.

## [editorinit](http://localhost:8125/ckeditor5/tests/manual/performance/editorinit.html) - long, semantic content

Content stats: 6174 elements, 9.228 text nodes.

tl;dr; `src/model/position#get parent` (17.1 self %) and `getRawTag` (16.9 self %) hit the town.

Click handling took 19.4 s.

As the content gets larger you can see that, intuitively, model and editable output takes more time.

One thing to note is that `src/model/position#get parent` function is a relatively small and fast one, but it's called a staggering **858k** times.

![](https://user-images.githubusercontent.com/5353898/73743926-f4024780-474f-11ea-9048-db8f1c9a09aa.png)

## [editorinit](http://localhost:8125/ckeditor5/tests/manual/performance/editorinit.html) - short, styled content

Click handling took 1.48s.

![](https://user-images.githubusercontent.com/5353898/73750106-e7d0b700-475c-11ea-88bb-8b3a7db803e9.png)

![](https://user-images.githubusercontent.com/5353898/73750189-06cf4900-475d-11ea-827f-d73770f91f20.png)

In this case time distribution is almost 50%/50% - half going for editor UI view initialization and another half on editable handling (toView, toModel, etc.).

## [editorinit](http://localhost:8125/ckeditor5/tests/manual/performance/editorinit.html) - full website styled content

Click handling took 6.28s.

![](https://user-images.githubusercontent.com/5353898/73748118-1c427400-4759-11ea-8a7d-15b11c087623.png)

![](https://user-images.githubusercontent.com/5353898/73748230-4300aa80-4759-11ea-9f5b-06436e86283c.png)

Call graph goes much deeper than previous ones, probably because there's more nested content.

Again, lodash's `getRawTag` and `src/model/position#get parent` function came first when looking at the self execution time.

View/model generation takes the most of the time.

## Conditions

Tests were performed on latest `master` branch today, in Chrome 79.0.3945.130 running on Windows 10.

**Comment by mlewand** [comment]
There are some quick wins possible. I recall seeing `getRawTag` (which is a result of calling `isPlainObject`) as a performance issue longer time ago.

Time given in seconds.

case | master | i/4504-rebased branch | improvement
---------|---------|----------|---------
medium semantic content | 5.94 | 4.62 | 22.22%
full styled website |  6.28 | 5.11 | 18.63%
short styled content |  1.48 | 1.22 | 15.27%

Pushed the changes to `i/4504-rebased` engine and utils repos.

https://github.com/ckeditor/ckeditor5-utils/tree/i/4504c_rebased
https://github.com/ckeditor/ckeditor5-engine/tree/i/4504-rebased

**Comment by oleq** [comment]
Related https://github.com/ckeditor/ckeditor5-utils/pull/322.

**Comment by mlewand** [comment]
There are couple of issues created from this one:

*   [#6195](https://github.com/ckeditor/ckeditor5/issues/6195): Improve listenTo() and fire() function execution time
*   [#6194](https://github.com/ckeditor/ckeditor5/issues/6194): Toolbar grouping should recalculate the styles less frequently
*   [#6193](https://github.com/ckeditor/ckeditor5/issues/6193): Table's InsertTableView takes a long time to render
*   [#6192](https://github.com/ckeditor/ckeditor5/issues/6192): Font color/background UI rendering should be improved

Also couple other things could be noted based on research but these probably couldn't be fixed with reasonable effort:

*   SVG parsing ([DomParser().parseFromString()](https://github.com/ckeditor/ckeditor5-ui/blob/ec39406/src/icon/iconview.js#L98)) takes a little of time in case of initializing editor with small amount of data.
*   `view.(element/node)` / `model.(element/node)` `is` regexps usage could be optimized (reuse regexps or try to use less performance hungry functions, e.g. String.startsWith / String.endsWith). It starts to take 2-3% of total time in case of medium and longer, semantic content. But is irrelevant for smaller content cases.

**Comment by oleq** [comment]
> * SVG parsing ([DomParser().parseFromString()](https://github.com/ckeditor/ckeditor5-ui/blob/ec39406/src/icon/iconview.js#L98)) takes a little of time in case of initializing editor with small amount of data.

It ~10ms startup time when switched to `innerHTML` in an editor with an average toolbar (8 icons). AFAIR this caused some problems in Edge but since Edge will be officially Chromium soon, we may consider this small tweak (works in Chrome, FF and Safari).
